src/compiler/crystal/interpreter/cast.cr

```diff
@@ -59,6 +59,18 @@ class Crystal::Repl::Compiler
   end
 
   private def upcast_distinct(node : ASTNode, from : PrimitiveType | EnumType | NonGenericClassType | GenericClassInstanceType | GenericClassInstanceMetaclassType | MetaclassType, to : MixedUnionType)
+    # It might happen that `from` is not of the union but it's compatible with one of them.
+    # We need to first cast the value to the compatible type and to `to`.
+    # This same logic exists in codegen/cast.cr
+    case from
+    when TupleInstanceType, NamedTupleInstanceType
+      unless to.union_types.any? &.==(from)
+        compatible_type = to.union_types.find { |ut| from.implements?(ut) }.not_nil!
```

Comment:
There's `Enumerable#find!` now :)

```suggestion
        compatible_type = to.union_types.find! { |ut| from.implements?(ut) }
```

Reply:
Thanks! I didn't know that was already in master.